instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.webmvc;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpServerMetrics;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpSpanStatusExtractor;
+import io.opentelemetry.instrumentation.servlet.javax.JavaxHttpServletRequestGetter;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** A builder of {@link SpringWebMvcTracing}. */
+public final class SpringWebMvcTracingBuilder {
+
+  private static final String INSTRUMENTATION_NAME = "io.opentelemetry.spring-webmvc-3.1";
+
+  private final OpenTelemetry openTelemetry;
+  private final List<AttributesExtractor<HttpServletRequest, HttpServletResponse>>
+      additionalExtractors = new ArrayList<>();
+
+  SpringWebMvcTracingBuilder(OpenTelemetry openTelemetry) {
+    this.openTelemetry = openTelemetry;
+  }
+
+  /**
+   * Adds an additional {@link AttributesExtractor} to invoke to set attributes to instrumented
+   * items.
+   */
+  public SpringWebMvcTracingBuilder addAttributesExtractor(
+      AttributesExtractor<HttpServletRequest, HttpServletResponse> attributesExtractor) {
+    additionalExtractors.add(attributesExtractor);
+    return this;
+  }
+
+  /**
+   * Returns a new {@link SpringWebMvcTracing} with the settings of this {@link
+   * SpringWebMvcTracingBuilder}.
+   */
+  public SpringWebMvcTracing build() {
+    SpringWebMvcHttpAttributesExtractor httpAttributesExtractor =
+        new SpringWebMvcHttpAttributesExtractor();
+
+    Instrumenter<HttpServletRequest, HttpServletResponse> instrumenter =
+        Instrumenter.<HttpServletRequest, HttpServletResponse>newBuilder(
+                openTelemetry,
+                INSTRUMENTATION_NAME,
+                HttpSpanNameExtractor.create(httpAttributesExtractor))
```

Comment:
If I understand correctly this just sets span name to `HTTP GET`, is it correct to say that this fixes #3479? I guess we could somehow adapt `OpenTelemetryHandlerMappingFilter` for library instrumentation that should give better names.

Reply:
> If I understand correctly this just sets span name to `HTTP GET`, is it correct to say that this fixes #3479?

🤷‍♂️  It's better than the class name I guess, and it fits the HTTP spec.
We can't really get better names on the `Filter` level, we'd probably have to implement some additional piece of code that injects itself into Spring machinery - I'm not really sure whether this is worth the effort though.

Reply:
Although it is out of the scope of this pr we already have this beast https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java A filter that deduces which spring mapping is going to match and sets the name of the server span. I guess the intention behind it is that if request is denied by spring security then it doesn't reach to actual matching so to get a good span name we need to figure out what matches the request before the actual matching in spring takes place. The same approach could probably also be used in library instrumentation.

Reply:
That's wild, but I think it's doable. I'll remove the association with #3479 and link your comment there - it's a good idea to at least try it out. Thanks!